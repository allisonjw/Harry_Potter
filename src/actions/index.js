export const addCharacters = characters => ({
    type: 'ADD_CHARACTERS',
    characters
});

export const addHouses = houses => ({
    type: 'ADD_HOUSES',
    houses
});

export const addSpells = spells => ({
    type: 'ADD_SPELLS',
    spells
});

export const isLoading = isLoading => ({
    type: 'SET_LOADING',
    isLoading
});

export const addHouseCharacters = filteredStudents => ({
    type: 'SET_CHARACTERS',
    filteredStudents
});

export const addFavorite = (id) => ({
    type: 'ADD_FAVORITE',
    id
});
  
export const deleteFavorite = (id) => ({
    type: 'DELETE_FAVORITE',
    id
});