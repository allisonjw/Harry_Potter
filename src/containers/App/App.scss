@import url("https://fonts.googleapis.com/css?family=Henny+Penny|ZCOOL+XiaoWei&display=swap");

* {
    margin: 0;
    font-family: 'Henny Penny', cursive;
}