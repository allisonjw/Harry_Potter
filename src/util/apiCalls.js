const apiKey = `?key=$2a$10$UN/Up5FVuauhCHKM/olg7efLQBTNWpkOjodpao2ICW0zo5V9zKArq`

export const getPotterApi = async path => {
    const url = `https://www.potterapi.com/v1/${path}${apiKey}`;
    const response = await fetch(url);
    if (!response.ok) {
      throw Error('Error fetching Potter Api');
    }
    const data = await response.json();
    return data
};

export const getCharacterHouse = async (url) => {
    const url2 = 'https://www.potterapi.com/v1/sortinghat'
    const response = await fetch(url2)
    if (!response.ok) {
      throw Error('Error fetching character house');
    }
    const data = await response.json();
    return data;
};