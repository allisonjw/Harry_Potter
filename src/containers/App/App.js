import React, { Component } from 'react';
import { getPotterApi } from '../../util/apiCalls';
import { addCharacters, addHouses, addSpells, isLoading } from '../../actions';
import Login from '../Login/Login';
import Navbar from '../../components/Navbar/Navbar';
import Charms from '../../components/Charms/Charms';
import CharacterContainer from '../CharacterContainer/CharacterContainer';
import HousesContainer from '../HousesContainer/HousesContainer';
import Favorites from '../Favorites/Favorite';
import { Route, Switch } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import './App.scss';
import { connect } from 'react-redux';


export class App extends Component {

    componentDidMount = async() => {
        const { addCharacters, addHouses, addSpells, isLoading } = this.props;
        try {
          isLoading(true);
          const characters = await getPotterApi('characters');
          addCharacters(characters);
          const houses = await getPotterApi('houses');
          addHouses(houses);
          const spells = await getPotterApi('spells');
          addSpells(spells);
          isLoading(false)
        } catch({ message }) {
          isLoading(false)
          console.log(message);
        }
     };
    
    render() {
        return (
            <Switch>
               <Route exact path='/login' render={() => <Login />} />
               <Route path='/houses' render={() => <> <Navbar /> <HousesContainer /> </>}/>
               <Route path='/charms' render={() => <> <Navbar /> <Charms /> </>}/>
               <Route path='/character' render={() => <> <Navbar /> <CharacterContainer /> </>}/>
               <Route path='/favorites' render={() => <> <Navbar /> <Favorites /> </>}/>
            </Switch>
        )
    }
};

export const mapStateToProps = state => ({
    allCharacters: state.allCharacters,
    allHouses: state.allHouses,
    allSpells: state.allSpells,
    isLoading: state.isLoading
})

export const mapDispatchToProps = dispatch => {
    return bindActionCreators({ addCharacters, addHouses, addSpells, isLoading}, dispatch) 
};

export default connect(null, mapDispatchToProps)(App)
