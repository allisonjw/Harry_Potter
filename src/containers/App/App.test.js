import React from 'react';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { getPotterApi } from '../../util/apiCalls';
import * as actions from '../../actions/index';

jest.mock('../../util/apiCalls.js');
jest.mock('../../actions');

describe('App', () => {
  let allCharacters = [{}, {}, {}]
  let allHouses = [{}, {}, {}]
  let allSpells = [{}, {}, {}]
  let wrapper = shallow(<App 
    addCharacters={actions.addCharacters}
    addHouses={actions.addHouses}
    addSpells={actions.addSpells}
    isLoading={actions.isLoading}
    allCharacters={allCharacters}
    allHouses={allHouses}
    allSpells={allSpells}
  />);

 it('should match the snapshot', () => {
  expect(wrapper).toMatchSnapshot();
 });

  describe('componentDidMount', () => {
    it('should call addCharacters', () => {
    expect(actions.addCharacters).toHaveBeenCalled();
    });

    it('should call addHouses', () => {
      expect (actions.addHouses).toHaveBeenCalled();
    });

    it('should call addSpells', () => {
      expect(actions.addSpells).toHaveBeenCalled();
    });

    it('should call isLoading', () => {
      expect(actions.isLoading).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with 4 paths', () => {
      let allCharacters, allHouses, allSpells, isLoading;
      let mockState = {
        allCharacters,
        allHouses,
        allSpells,
        isLoading
      };
  
      let expected = {
        allCharacters,
        allHouses,
        allSpells,
        isLoading
      };
  
      const mappedProps = mapStateToProps(mockState);
  
      expect(mappedProps).toEqual(expected);

    });
  });

  describe('mapDispatchToProps', () => {
    let mockDispatch;
  
    beforeEach(() => {
      mockDispatch = jest.fn();
    });
  
    it('should call dispatch with a addCharacters action', () => {
      const actionToDispatch = actions.addCharacters([{}, {}, {}]);
      const mappedProps = mapDispatchToProps(mockDispatch);
  
      mappedProps.addCharacters([{}, {}, {}]);
  
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with a addHouses action', () => {
      const actionToDispatch = actions.addHouses([{}, {}, {}]);
      const mappedProps = mapDispatchToProps(mockDispatch);
  
      mappedProps.addHouses([{}, {}, {}]);
  
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with a addSpells action', () => {
      const actionToDispatch = actions.addSpells([{}, {}, {}]);
      const mappedProps = mapDispatchToProps(mockDispatch);
  
      mappedProps.addSpells([{}, {}, {}]);
  
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with a isLoading action', () => {
      const actionToDispatch = actions.isLoading(true);
      const mappedProps = mapDispatchToProps(mockDispatch);
  
      mappedProps.isLoading(true);
  
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});    