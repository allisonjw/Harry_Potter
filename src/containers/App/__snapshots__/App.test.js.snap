// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<Switch>
  <Route
    exact={true}
    path="/login"
    render={[Function]}
  />
  <Route
    path="/houses"
    render={[Function]}
  />
  <Route
    path="/CharmsContainer"
    render={[Function]}
  />
  <Route
    path="/character"
    render={[Function]}
  />
  <Route
    path="/FavoriteContainer"
    render={[Function]}
  />
  <Redirect
    to="/login"
  />
</Switch>
`;
